File: awk/Cpp/awkLexerBase.cpp
#include "awkLexerBase.h"
#include "awkLexer.h"

std::unique_ptr<antlr4::Token> awkLexerBase::nextToken()
{
    std::unique_ptr<antlr4::Token> next = antlr4::Lexer::nextToken();

    if (next->getChannel() == antlr4::Token::DEFAULT_CHANNEL) {
        // Keep track of the last token on the default channel.
	    this->_afterExpr = next->getType() == awkLexer::WORD
			      || next->getType() == awkLexer::NUMBER
			      || next->getType() == awkLexer::STRING
			      || next->getType() == awkLexer::BUILTIN_FUNC_NAME
			      || next->getType() == awkLexer::INCR
			      || next->getType() == awkLexer::DECR
			      || next->getType() == awkLexer::Rp
			      || next->getType() == awkLexer::Rb;
    }

    return next;
}

bool awkLexerBase::IsNotAfterExpr()
{
    return ! this->_afterExpr;
}

File: awk/Dart/awkLexerBase.dart
import 'package:antlr4/antlr4.dart';
import 'awkLexer.dart';

abstract class awkLexerBase extends Lexer {
  bool _afterExpr = false;

  awkLexerBase(CharStream input) : super(input);

  @override
  Token nextToken() {
    final token = super.nextToken();
    if (token.channel == Token.DEFAULT_CHANNEL) {
      _afterExpr = token.type == awkLexer.TOKEN_WORD
          || token.type == awkLexer.TOKEN_NUMBER
          || token.type == awkLexer.TOKEN_STRING
          || token.type == awkLexer.TOKEN_BUILTIN_FUNC_NAME
          || token.type == awkLexer.TOKEN_INCR
          || token.type == awkLexer.TOKEN_DECR
          || token.type == awkLexer.TOKEN_Rp
          || token.type == awkLexer.TOKEN_Rb;
    }
    return token;
  }

  bool IsNotAfterExpr() => !_afterExpr;
}
